### ISSUE TITLE
SVT: [Remote] Performance Dashboard UI can present the load test stuck in "Requested"

### ISSUE THREAD

**Issue by sujeilyfonseca** [eclipse/codewind#2284]
**Codewind version:** 0.9.0
**OS:** VM with Windows 10 (w. OKD/Openshift cluster)

**IDE extension version:** 0.9.0
**IDE version:** 1.42.0

**Description:** @jagraj and I noticed that the Performance Dashboard UI can present the load test stuck in "Requested". If users try to cancel the request, they will not be able to do so:

<img width="1671" alt="Screen Shot 2020-02-20 at 4 58 04 PM" src="https://user-images.githubusercontent.com/49531960/74986220-1e6a2b00-5407-11ea-92fc-05445cbd71b4.png">

However, the load test runs successfully. Users can see the test results, if they open a new window:

<img width="1655" alt="Screen Shot 2020-02-20 at 5 11 37 PM" src="https://user-images.githubusercontent.com/49531960/74986131-f7135e00-5406-11ea-90a1-b5a323b8cd90.png">

Also, if they attempt to run new load tests in the new window, they will not see the problem again:

<img width="1617" alt="Screen Shot 2020-02-20 at 5 32 51 PM" src="https://user-images.githubusercontent.com/49531960/74986197-1316ff80-5407-11ea-90f4-7c7e0b77c15a.png">

```
[root@sfonseca-okd-codewind ~]# oc  logs  codewind-performance-k6v9sv51-84f57c6cd7-hkgmw 

> performanceserver@1.0.0 start /usr/src/app
> PORT=9095 node server.js

[20/02/20 21:42:36 /usr/src/app/server.js] [INFO] Performance server listening on port 9095!
[20/02/20 21:58:56 /usr/src/app/server.js] [INFO] starting loadrun with options : {"path":"/","requestsPerSecond":"1","concurrency":"1","maxSeconds":"180","url":"http://172.30.209.241:9080/"}
[20/02/20 22:01:57 /usr/src/app/server.js] [INFO] completed - loadrun summary : { totalRequests: 180,
  totalErrors: 0,
  totalTimeSeconds: 180.005692361,
  rps: 1,
  meanLatencyMs: 6.1,
  maxLatencyMs: 67,
  minLatencyMs: 4,
  percentiles: { '50': 5, '90': 6, '95': 7, '99': 11 },
  errorCodes: {},
  instanceIndex: 0 }

[20/02/20 22:12:49 /usr/src/app/server.js] [INFO] starting loadrun with options : {"path":"/","requestsPerSecond":"1","concurrency":"1","maxSeconds":"180","url":"http://172.30.209.241:9080/"}
[20/02/20 22:15:49 /usr/src/app/server.js] [INFO] completed - loadrun summary : { totalRequests: 180,
  totalErrors: 0,
  totalTimeSeconds: 180.005418932,
  rps: 1,
  meanLatencyMs: 6,
  maxLatencyMs: 30,
  minLatencyMs: 4,
  percentiles: { '50': 5, '90': 6, '95': 7, '99': 9 },
  errorCodes: {},
  instanceIndex: 0 }

[20/02/20 22:19:01 /usr/src/app/server.js] [INFO] starting loadrun with options : {"path":"/","requestsPerSecond":"1","concurrency":"1","maxSeconds":"180","url":"http://172.30.209.241:9080/"}
[20/02/20 22:22:01 /usr/src/app/server.js] [INFO] completed - loadrun summary : { totalRequests: 180,
  totalErrors: 0,
  totalTimeSeconds: 180.00626806100001,
  rps: 1,
  meanLatencyMs: 5.7,
  maxLatencyMs: 29,
  minLatencyMs: 4,
  percentiles: { '50': 5, '90': 6, '95': 7, '99': 10 },
  errorCodes: {},
  instanceIndex: 0 }
```

**Workaround:** 
1. If, after executing a load test, users see the status stuck in "Requested", they should open a new Performance Dashboard window.

@jagraj

**Comment by jagraj** [comment]
I also got into similar issue with the hybrid scenario. The load test status stuck at "Requested" state and I need to close the window and open new window to submit another request or wait for first request to complete. Here are my screenshots. We do not see same issue in Eclipse Che and this issue can be addressed in next  release as we have workaround.

![Screen Shot 2020-02-20 at 5 47 27 PM](https://user-images.githubusercontent.com/713849/74988031-aef63a80-540a-11ea-826f-314e304a0302.png)
![Screen Shot 2020-02-20 at 5 49 28 PM](https://user-images.githubusercontent.com/713849/74988052-c3d2ce00-540a-11ea-92fb-41f6cdef66a9.png)
![Screen Shot 2020-02-20 at 5 51 39 PM](https://user-images.githubusercontent.com/713849/74988071-d2b98080-540a-11ea-91ac-af7df6db8261.png)

**With the new window.**
![Screen Shot 2020-02-20 at 5 52 19 PM](https://user-images.githubusercontent.com/713849/74988075-d5b47100-540a-11ea-952e-9140f0c6261a.png)

**Comment by markcor11** [comment]
Yes,  we know about this one, the cancel method is being overhauled at the moment.  The changes that went in recently for 0.9.0 switch the status to "Requesting" earlier in the cycle.  This is to protect the user from `accidentally` clicking the button to start a second test before the first one completes. 

The status change from `Requested` to `Preparing` is controlled via a socket message from PFE back to the dashboard.  That same socket pipe is also used to control the flow from `Cancelling` to `Cancelled`. However,  if the browser lost socket connection to PFE and didn't manage to reconnect it,  those socket events will never be heard. The REST requests to start load and cancel are performed over REST and will fire (which is why the load run did start and did complete) but in your case,  the UI didn't get back the status changes to refresh the button label because it likely lost that critical socket session.

We are going to harden this flow a few ways : 

1.  Ensure that the UI knows about socket connected and disconnected state and display that on the dashboard with a "connected" and "disconnected" icon.
2. Block all interaction with the UI when socket status is disconnected forcing a page reload and protect against any REST/Socket sync issues
3. Historically "isLoadRunning" is a boolean and it's just not refined enough to hold state about the many recent additions in Codewind like app profiling, load flow control, concurrency etc.  We'll replace this boolean with a phase controller object that manages the entire workflow of all parts of running a load experiment.

**Comment by markcor11** [comment]
The cancel and run methods are being overhauled at the moment.  The changes that went in recently for 0.9.0 switch the status to "Requesting" earlier in the cycle.  This is to protect the user from `accidentally` clicking the button to start a second test before the first one completes. 

The status change from `Requested` to `Preparing` are controlled via a socket message from PFE back to the dashboard.  That same socket pipe is also used to control the flow from `Cancelling` to `Cancelled`. However,  if the browser lost socket connection to PFE and didn't manage to reconnect it,  those socket events will never be heard. The dashboard action to start load and cancel are performed over REST, they will fire (which is why the load run did start and did complete),  but in your case the UI didn't get back the socket status change to refresh the button label because the socket connection had dropped some time earlier. 

We are going to harden this flow a few ways : 

1.  Ensure that the UI knows about socket connected and disconnected state and display that on the dashboard with a "connected" and "disconnected" icon.
2. Block all interaction with the UI when socket status is disconnected forcing a page reload and protect against any REST/Socket sync issues
3. Historically "isLoadRunning" is a boolean and it's just not refined enough to hold state about the many recent additions in Codewind like app profiling, load flow control, concurrency etc.  We'll replace this boolean with a phase controller object that manages the entire workflow of all parts of running a load experiment.

**Comment by malincoln** [comment]
@tobespc are you targeting 0.10.0 for fix?

**Comment by tobespc** [comment]
this has been fixed already, can not reproduce with the fixes in

**Comment by malincoln** [comment]
@sujeilyfonseca pls verify and close. Thanks

**Comment by sujeilyfonseca** [comment]
I've verified this issue with the latest Codewind 0.10.0, and I no longer reproduce the problem.

**Comment by sujeilyfonseca** [comment]
I reopened this problem as we are seeing this problem again with Codewind 0.10.0:

<img width="624" alt="Screen Shot 2020-03-18 at 6 24 35 PM" src="https://user-images.githubusercontent.com/49531960/77013628-cd3a5200-6946-11ea-86a0-3e7c7412843a.png">

<img width="1518" alt="Screen Shot 2020-03-18 at 6 31 12 PM" src="https://user-images.githubusercontent.com/49531960/77013656-dd523180-6946-11ea-8ad6-8ab7a47312cb.png">

**Comment by jagraj** [comment]
I also reproduced same problem with the remote/hybrid scenario. After clicking "Run Load Test" , the status get stuck into "Requested". If we wait for few minutes and refresh the page then we can see the results. 

![image (3)](https://user-images.githubusercontent.com/713849/77014172-1212b880-6948-11ea-8abf-78ed4c1a2200.png)

After refresh..


![image](https://user-images.githubusercontent.com/713849/77014243-37072b80-6948-11ea-9820-01e3330efebb.png)

**Comment by jagraj** [comment]
I just tried to reproduce this problem again for failed project and now the request is not hung.Looks like this is intermittent problem. Here are the logs from my environment which should have for both failed and successful load test. 

[performance.txt](https://github.com/eclipse/codewind/files/4354732/performance.txt)
[pfe.txt](https://github.com/eclipse/codewind/files/4354733/pfe.txt)

**Comment by malincoln** [comment]
@jagraj is this considered stopship?  I don't see the label.  Also, if it's intermittent, maybe we keep it hot and add "intermittent" in title?

**Comment by malincoln** [comment]
Moving from verify to In progress and assigning to @tobespc

**Comment by malincoln** [comment]
Decided during discussion to continue to investigate this for 0.10.0 and see if we have a fix by mid next week and decide whether we ship with/without fix.

**Comment by sujeilyfonseca** [comment]
As a general note about this issue, it can be reproduced more often with new Codewind installations. If Codewind has been running for a while, the occurrence of this issue may be reduced.

Regarding environments, I have seen this issue more often in remote scenarios. However, it may be reproduced in Eclipse Che.

Once you hit this issue, there are three possible scenarios:

1. Cancel will not work, and you will see the UI with a "Cancelling" status. If you refresh the page or open a new window, you will not be able to run load tests with the same project:

<img width="1377" alt="Screen Shot 2020-03-20 at 10 44 16 AM" src="https://user-images.githubusercontent.com/49531960/77180810-d6d7cd00-6aa0-11ea-9d5b-4e31ca9584e1.png">

<img width="694" alt="Screen Shot 2020-03-20 at 10 44 32 AM" src="https://user-images.githubusercontent.com/49531960/77179077-5e700c80-6a9e-11ea-8f4e-634f06b04e89.png">

<img width="1374" alt="Screen Shot 2020-03-20 at 10 48 33 AM" src="https://user-images.githubusercontent.com/49531960/77179082-616afd00-6a9e-11ea-8958-1be76e7c37c3.png">

2. After refreshing the page and waiting some minutes, you can get test results:

<img width="1373" alt="Screen Shot 2020-03-20 at 11 37 57 AM" src="https://user-images.githubusercontent.com/49531960/77179747-43ea6300-6a9f-11ea-8ab5-b08339caa576.png">

3. After refreshing the page and waiting some minutes, you will not see anything and will see the same if you try to run another test:

<img width="578" alt="Screen Shot 2020-03-20 at 11 36 25 AM" src="https://user-images.githubusercontent.com/49531960/77179583-0be32000-6a9f-11ea-8d0a-edb3dd7072aa.png">

![Screen Shot 2020-03-20 at 11 32 48 AM](https://user-images.githubusercontent.com/49531960/77179548-01c12180-6a9f-11ea-906e-e0eda175c47c.png)

Here are some additional logs to help solve this problem:

[remote-codewind-pfe.txt](https://github.com/eclipse/codewind/files/4360514/remote-codewind-pfe.txt)
[remote-codewind-performance.txt](https://github.com/eclipse/codewind/files/4360513/remote-codewind-performance.txt)

[che-codewind-pfe.txt](https://github.com/eclipse/codewind/files/4360506/che-codewind-pfe.txt)
[che-codewind-performance.txt](https://github.com/eclipse/codewind/files/4360509/che-codewind-performance.txt)

**Comment by jagraj** [comment]
After installing remote codewind, I created Appsody Node.js Scaffold, Appsody Eclipse Microprofile, Default Node.js express, WebSphere liberty projects. When the projects start running, I tried to access "Performance dashboard" and started "Run Load Test" for "Default Node.js Express" project. The status goes into "Requested" state and when we try to "Cancel" it then it get stuck at "Canceling" state.

Here are the logs and screenshots from the failure. 

![Screen Shot 2020-03-20 at 10 50 02 AM](https://user-images.githubusercontent.com/713849/77180210-fa4e4800-6a9f-11ea-9ad7-1ed5650d1caa.png)



Browser console errors.
```
bundle.js:36 [KEYCLOAK] Using legacy promises is deprecated and will be removed in future versions. You can opt in to using native promises by setting `promiseType` to 'native' when initializing Keycloak.
a.init @ bundle.js:36
bundle.js:63 WebSocket connection to 'wss://codewind-gatekeeper-k809w0yq.apps.extols.os.fyre.ibm.com/socket.io/?EIO=3&transport=websocket&sid=-i9urRqOuVlWvM_jAAAE' failed: Error during WebSocket handshake: Unexpected response code: 400
d.doOpen @ bundle.js:63
bundle.js:79 Dashboard - authenticated
main.js:1 Result card - No available CPU data
main.js:1 Result card - No available memory data
main.js:1 Result card - No available HTTP response data
bundle.js:79 Dashboard refresh-token refreshed
(anonymous) @ bundle.js:79
DevTools failed to parse SourceMap: chrome-extension://mbopgmdnpcbohhpnfglgohlbhfongabi/sideex/browser-polyfill.js.map
```

**Comment by micgibso** [comment]
New Troubleshooting issue drafted on above PR.

**Comment by jagraj** [comment]
@tobespc @micgibso 
Based on the comments provided in this issue  (https://github.com/eclipse/codewind/issues/2284#issuecomment-601770507) the workaround solution will not help to the users always. Our understanding on Friday checkpoint call that we are planning to deliver fix for this issue in Codewind 0.10.0 release.

**Comment by tobespc** [comment]
As I understand it, refreshing the browser does solve the issue.   Its the ui socket disconnecting that causes the issue and if we can detect that, we can output a message to the user informing them.

To change this in code is a large change and not something I believe we should do under a stop ship.  If we have a suitable workaround then I would propose we use that and fix this in a different way for 0.11.0

But, do we agree on the workaround ?  Its hard to tell from this thread what the current status actually is so be worth discussing it in a call

**Comment by micgibso** [comment]
Docs no longer required, programmatically rectified.

**Comment by malincoln** [comment]
> Decided during discussion to continue to investigate this for 0.10.0 and see if we have a fix by mid next week and decide whether we ship with/without fix.

Here is the update after the call on Friday.  There were more updates from SVT after my update

**Comment by sujeilyfonseca** [comment]
I cleaned up my environments and executed tests immediately after getting the latest builds. I covered scenarios with Eclipse Che and VS Code (Remote). Here are my findings:

**1. Eclipse Che (Fyre OCP 4.3 Cluster):** 
I deployed a new Codewind 0.10.0 workspace and proceeded to create four Default (Codewind-style) projects. Then, I continued to wait until at least one of them was in "Running" state:

```
[root@cw-codewind-43-inf ~]# oc get pods
NAME                                                              READY   STATUS      RESTARTS   AGE
che-65ddf6b565-thwwh                                              1/1     Running     0          6d19h
che-operator-5fc5d6d8d7-98xhd                                     1/1     Running     0          6d19h
codewind-performance-workspace7wd59uozf2kus44z-5846fccdf7-22khk   1/1     Running     0          78m
codewind-workspace7wd59uozf2kus44z-b588fd594-vskjr                1/1     Running     0          78m
cw-cwchemicroprofile010-ef93d810-6d26-11ea-9773-648b4dffdfjp557   1/1     Running     0          65m
cw-cwchenode010-089201c0-6d27-11ea-9773-c987d8b87-4z7sx           1/1     Running     0          61m
cw-cwchespring010-03ed31d0-6d27-11ea-9773-6cb9f96688-m9gnw        1/1     Running     0          66m
cw-cwcheswift010-f89ac9a0-6d26-11ea-9773-75f48cf8b5-5gxz7         1/1     Running     0          65m
devfile-registry-7d9d8b8dd4-smms8                                 1/1     Running     0          6d19h
keycloak-66f6d6c444-nbv6f                                         1/1     Running     0          6d19h
plugin-registry-6d94b964db-zbcwx                                  1/1     Running     0          6d19h
postgres-57c4df88dd-rhs7p                                         1/1     Running     0          6d19h
workspace7wd59uozf2kus44z.che-workspace-pod-7fdf86fd5d-fcflr      6/6     Running     0          79m
```

<img width="1676" alt="Screen Shot 2020-03-23 at 1 03 36 PM" src="https://user-images.githubusercontent.com/49531960/77348434-d42edf00-6d0f-11ea-9a81-101f42c82caf.png">

After that, I executed a load test with a Swift project, and it was successful. I noticed the UI presenting a "Connected" status, which is expected with the latest builds.

<img width="1663" alt="Screen Shot 2020-03-23 at 1 07 37 PM" src="https://user-images.githubusercontent.com/49531960/77348575-1526f380-6d10-11ea-8386-1708910cebee.png">

Then, I ran a load test with a Spring project, and it seemed to start successfully.

[Screen Recording 2020-03-23 at 1.07.55 PM.mov.zip](https://github.com/eclipse/codewind/files/4371170/Screen.Recording.2020-03-23.at.1.07.55.PM.mov.zip)

At this time, I wanted to test if a new test with another project will trigger that a load test is currently in progress. I wanted to check that because before, I didn't notice the "A load test is in progress" message, and I informed Toby about that.

I proceeded to run a load test with a Node JS project. The UI was in a "Connected" status, but I didn't get a message indicating that I still have a Spring test in progress. Therefore, my Node JS load test got stuck in "Requested". After waiting, my Spring test was successful. 

<img width="1659" alt="Screen Shot 2020-03-23 at 1 11 44 PM" src="https://user-images.githubusercontent.com/49531960/77349144-f2490f00-6d10-11ea-8d1a-ec899bd90cf7.png">

Then, I executed a load test with a MicroProfile project. It seemed to start successfully, but I didn't get any results.

[Screen Recording 2020-03-23 at 1.11.53 PM.mov.zip](https://github.com/eclipse/codewind/files/4371235/Screen.Recording.2020-03-23.at.1.11.53.PM.mov.zip)

<img width="1660" alt="Screen Shot 2020-03-23 at 1 17 16 PM" src="https://user-images.githubusercontent.com/49531960/77350176-7cde3e00-6d12-11ea-8e88-5481c75ffd7b.png">

Next, I started another test with the same MicroProfile project. The UI status was "Connected", but my test got stuck in "Requested".

[MicroProfile-Test-2.mov.zip](https://github.com/eclipse/codewind/files/4371250/MicroProfile-Test-2.mov.zip)

<img width="1672" alt="Screen Shot 2020-03-23 at 1 20 57 PM" src="https://user-images.githubusercontent.com/49531960/77350399-e0686b80-6d12-11ea-9844-8928ae2f3eb6.png">

After waiting a while, I closed all my open tabs but still couldn't run the load test with the same MicroProfile project—however, this time, I received the error message saying that another test was in progress.

[Screen Recording 2020-03-23 at 1.56.35 PM.mov.zip](https://github.com/eclipse/codewind/files/4371264/Screen.Recording.2020-03-23.at.1.56.35.PM.mov.zip)

 It is important to notice that the UI status has always been "Connected".

<img width="92" alt="Screen Shot 2020-03-23 at 3 01 38 PM" src="https://user-images.githubusercontent.com/49531960/77353355-3808d600-6d17-11ea-94b3-dba045a1991d.png">

</br> 

**2. VS Code (Remote):**
I installed the latest builds of Codewind 0.10.0. After that, I created a new remote instance with the latest `cwctl`. Then, I waited until at least one of the projects was in "Running" state.

![Screen Shot 2020-03-23 at 1 11 01 PM](https://user-images.githubusercontent.com/49531960/77351087-b1062e80-6d13-11ea-8060-c8fb11bcac82.png)

![Screen Shot 2020-03-23 at 1 15 19 PM](https://user-images.githubusercontent.com/49531960/77351108-b3688880-6d13-11ea-8cd6-94af10f45a30.png)

Consequently, I tried to run a load test with a Node JS project, and the test got stuck in "Requested". Refreshing the browser didn't work either waiting for a long time before making another request.

[Remote-Node-Test-1_P1.mov.zip](https://github.com/eclipse/codewind/files/4371331/Remote-Node-Test-1_P1.mov.zip)

[Remote-Node-Test-1_P2.mov.zip](https://github.com/eclipse/codewind/files/4371334/Remote-Node-Test-1_P2.mov.zip)

![Screen Shot 2020-03-23 at 1 29 58 PM](https://user-images.githubusercontent.com/49531960/77352415-874e0700-6d15-11ea-8ad2-127d0df413b5.png)

After that, I ran a load test with a MicroProfile project and it was successful.

![Screen Shot 2020-03-23 at 2 51 29 PM](https://user-images.githubusercontent.com/49531960/77352620-d8f69180-6d15-11ea-85a5-d2d7c9c6064e.png)

Then, I executed a load test with a Swift project and it was successful.

[Screen Recording 2020-03-23 at 1.40.29 PM.mov.zip](https://github.com/eclipse/codewind/files/4371355/Screen.Recording.2020-03-23.at.1.40.29.PM.mov.zip)

![Screen Shot 2020-03-23 at 2 53 16 PM](https://user-images.githubusercontent.com/49531960/77352799-270b9500-6d16-11ea-9d64-3cc34b83877b.png)

Then, I tried a load test with the Spring project and it got stuck on "Requested". I couldn't retrieve the test results even after waiting a while.

[Screen Recording 2020-03-23 at 1.35.19 PM.mov.zip](https://github.com/eclipse/codewind/files/4371363/Screen.Recording.2020-03-23.at.1.35.19.PM.mov.zip)

![Screen Shot 2020-03-23 at 2 56 10 PM](https://user-images.githubusercontent.com/49531960/77353103-bfa21500-6d16-11ea-8457-74a68afee612.png)

It is important to notice that the UI status has always been "Offline".

[Screen Recording 2020-03-23 at 1.37.20 PM.mov.zip](https://github.com/eclipse/codewind/files/4371389/Screen.Recording.2020-03-23.at.1.37.20.PM.mov.zip)

</br>

**Logs:**
[che-codewind-workspace7wd59uozf2kus44z-b588fd594-vskjr.txt](https://github.com/eclipse/codewind/files/4371407/che-codewind-workspace7wd59uozf2kus44z-b588fd594-vskjr.txt)
[che-codewind-performance-workspace7wd59uozf2kus44z-5846fccdf7-22khk.txt](https://github.com/eclipse/codewind/files/4371406/che-codewind-performance-workspace7wd59uozf2kus44z-5846fccdf7-22khk.txt)

[vscode-remote-codewind-pfe-k84qbm0y-768dfbfcb5-xz5hv.txt](https://github.com/eclipse/codewind/files/4371420/vscode-remote-codewind-pfe-k84qbm0y-768dfbfcb5-xz5hv.txt)
[vscode-remote-codewind-performance-k84qbm0y-6c749fdf89-cbfh8.txt](https://github.com/eclipse/codewind/files/4371421/vscode-remote-codewind-performance-k84qbm0y-6c749fdf89-cbfh8.txt)

**Comment by tobespc** [comment]
Thanks for the update, will pick this up asap in the morning

**Comment by sujeilyfonseca** [comment]
Thanks, @tobespc!

**Comment by tobespc** [comment]
After more debugging, we are still not any closer but... I have some points for discussion
  * This is an error scenario that should be fixed
                   however
  * To recreate, I don't believe this is normal behaviour for a developer.   The recreation requires the environment of multiple microservices all being developed at the same time, all wanting to have load test applied to them. 
  * Whilst Codewind can support building and running multiple projects, that doesn't mean we have designed it to support actively developing multiple microservices concurrently.
  * From our experience, developers would be focused on coding and testing one thing at a time.  Other things might be running but would they want to be performance testing those at the same time ?
  * Don't think this is fixable in a short time window

**Comment by johnmcollier** [comment]
@elsony asked me to reproduce - As described in the issue, I can reproduce but only when loadtesting against multiple projects at once

**Comment by jgwest** [comment]
@elsony likewise asked me to reproduce... I wasn't able to reproduce using traditional methods, but I was able to "reproduce" using the following:
1) Create a Node project, wait for it to build and start.
2) Open the Performance Dashboard for the project.
3) Disconnect from VPN (or disconnect from WiFi). The 'Offline' status at the top right is displayed. Wait ~10 seconds. 
4) Reconnect to VPN(/WiFi). Immediately attempt to kick off a load test (don't wait for 'Connected' status).

You may see a TypeError dialog from Chrome, and the 'Requesting...' status will be displayed and never resolve. Refreshing the browser resolves the issue.

This simulates a network drop between the browser and the cluster; while I obviously wouldn't expect a user to reproduce using this specific set of steps, it might approximate what a user would encounter with a bad WiFi connection, for example.

For this I was testing with latest Codewind 0.10 on Che, within Chrome, on non-OpenShift Kube.

**Comment by jagraj** [comment]
> @elsony asked me to reproduce - As described in the issue, I can reproduce but only when loadtesting against multiple projects at once

@johnmcollier - Did you verify this scenario with Eclipse Che or Hybrid Scenario.?

**Comment by johnmcollier** [comment]
@jagraj Hybrid.

**Comment by sujeilyfonseca** [comment]
As you can see in the above scenarios where this problem has been reproduced, this can happen by accessing the Performance Dashboard using a single supported project.

https://github.com/eclipse/codewind/issues/2284#issuecomment-602805428
https://github.com/eclipse/codewind/issues/2284#issuecomment-601774669

**To summarize the steps that were performed:**
1. Install Codewind 0.10.0 
2. For remote scenarios: 
    - Use the recently installed 0.10.0 `cwctl` to deploy Codewind remotely
    - Create a new remote connection  
3. Create a new image registry
4. Create some projects (e.g., Node, Swift, MicroProfile, Spring)
5. Wait until at least one project is in "Running" state
6. Run a load test

**General notes:**
1. If your first load test was successful, try some additional tests as the problem may be intermittent, but more than one user has encountered it.
2. This problem has been reproduced in Eclipse Che, VS Code (Remote), and Eclipse (Remote). However, it can be reproduced more often in hybrid. 
3. For some reason, we can reproduce it immediately after the first attempt to run a load test in hybrid. However, it may happen after successfully retrieving load test results (see above comments).
4. Creating a load test when another test is being executed can trigger the issue. Currently, we are allowing users to open more than one Performance Dashboard tab and send more than one request. 
5. Losing the connection will trigger the issue.
6. 4 and 5 are ways to get the same problem, as noticed in previous comments. However, as mentioned before, it may even occur in your first attempt(s) to get load test results. You can also fail to get test results with a project that retrieved results before.
6. The new dashboard statuses were not always useful, as load test results were obtained with an "Offline" status, and the error also happened with a "Connected" status.

**Comment by markcor11** [comment]
Actually the new dashboard status is useful.  Offline indicates that the socket was not connected to PFE. It is over that socket channel that PFE drives the status button in the UI changing the state you see from Requested to Preparing, Running, Done etc.  

If the dashboard socket is not connected to PFE, PFE is unable to control the button status which leaves it in the last known state of "Requested". 

The action of changing the button from "Requesting" to "Requested" is performed over a REST request from Dashboard->PFE and not over the socket channel.

That is why the load run will actually complete because PFE did get told to start a load run (over REST), but the button still says "Requested" because the socket was offline and didn't get told to change to "Running" or "Completed" 

I do have a code change that will blocking access to the 'Run Load Test" button and all controls on the page until the socket reconnects which will make it clear "all bets are off".  However I have not activated that block yet whilst troubleshooting. More importantly is understanding why the socket won't re-connect.  My gut feeling at the moment is that the security token has expired in the browser and is not being refreshed by the UI when its idle.  To solve that, we may need to re-direct the user back through the login screen when the browser tab is bought back into focus.

**Comment by malincoln** [comment]
moving to verify

**Comment by sujeilyfonseca** [comment]
I retested this issue with the latest Codewind 0.10.0. I covered scenarios with Eclipse Che, using a Fyre OCP 4.3 cluster and a Google Chrome browser on Mac; and VS Code (Remote) using a Windows VM and a Firefox browser. Here are my findings:

**1. VS Code (Remote):** 
I installed the latest Codewind 0.10.0. After that, I created a new remote instance with the latest `cwctl`. Then, I created four default projects (MicroProfile, Swift, Spring, and Node) and waited for the projects to be in "Running" state.

<img width="1673" alt="Screen Shot 2020-03-26 at 12 24 47 PM" src="https://user-images.githubusercontent.com/49531960/77680353-371bb280-6f6a-11ea-9659-a8d8a2cd5d82.png">

Consequently, I tried to run a load test with a WebShere Liberty MicroProfile project. 

<img width="1656" alt="Screen Shot 2020-03-26 at 12 03 40 PM" src="https://user-images.githubusercontent.com/49531960/77680906-fc664a00-6f6a-11ea-827a-c89a198064bb.png">

After that, I opened another tab, and I was not allowed to do anything. This behavior is expected with the new modal window.

<img width="1659" alt="Screen Shot 2020-03-26 at 12 03 48 PM" src="https://user-images.githubusercontent.com/49531960/77680934-012afe00-6f6b-11ea-8cd6-a53547757e68.png">

I closed that second window and waited for my current load test in my first tab. My load test finished, but I didn’t get any results. Eventually, the "Run Load Test" button was enabled again, so I used it to run another test with the same project, and it got stuck in “Requested”.

<img width="1676" alt="Screen Shot 2020-03-26 at 12 07 43 PM" src="https://user-images.githubusercontent.com/49531960/77681014-20c22680-6f6b-11ea-8dff-dd953fc336ed.png">

From now on, I can't start any new request because each new window is blocked from doing anything due to an "Offline" status.

<img width="1629" alt="Screen Shot 2020-03-26 at 1 20 26 PM" src="https://user-images.githubusercontent.com/49531960/77681096-42bba900-6f6b-11ea-9439-265e2951e9a2.png">

</br>

**2. Eclipse Che:**
I deployed a new Codewind 0.10.0 workspace and proceeded to create four default projects (MicroProfile, Swift, Spring, and Node) and waited for the projects to be in "Running" state.

After that, I ran a load test with a WebShere Liberty MicroProfile project.

<img width="1674" alt="Screen Shot 2020-03-26 at 12 25 31 PM" src="https://user-images.githubusercontent.com/49531960/77681713-25d3a580-6f6c-11ea-82e6-227fa362e764.png">

Then, I opened another tab, and I was allowed to interact with it, but I didn't do anything and closed it.

<img width="1675" alt="Screen Shot 2020-03-26 at 12 26 29 PM" src="https://user-images.githubusercontent.com/49531960/77681843-574c7100-6f6c-11ea-9d8a-fb362babf007.png">

I waited for my current load test to finish, but I didn’t get any results. Eventually, the "Run Load Test" button was enabled again, so I used it to run another test with the same project, and it got stuck in “Requested”.

<img width="1676" alt="Screen Shot 2020-03-26 at 1 08 07 PM" src="https://user-images.githubusercontent.com/49531960/77681927-78ad5d00-6f6c-11ea-95a3-248042d808ae.png">

<img width="1673" alt="Screen Shot 2020-03-26 at 1 08 47 PM" src="https://user-images.githubusercontent.com/49531960/77681949-7fd46b00-6f6c-11ea-8efb-1201f42d0956.png">

From now on, I can try new load tests, but they get stuck in "Requested".

<img width="1677" alt="Screen Shot 2020-03-26 at 1 19 07 PM" src="https://user-images.githubusercontent.com/49531960/77681986-91b60e00-6f6c-11ea-9667-54922c50c596.png">

</br> 

**Logs:**
[vscode-remote-codewind-pfe-k88xliix-899d98d5-tjkcs.txt](https://github.com/eclipse/codewind/files/4389053/vscode-remote-codewind-pfe-k88xliix-899d98d5-tjkcs.txt)
[vscode-remote-codewind-performance-k88xliix-588c5d8859-lsdrm.txt](https://github.com/eclipse/codewind/files/4389054/vscode-remote-codewind-performance-k88xliix-588c5d8859-lsdrm.txt)

[che-odewind-workspace1zycwn4fbqw199tr-55b8788684-8znpr.txt](https://github.com/eclipse/codewind/files/4389079/che-odewind-workspace1zycwn4fbqw199tr-55b8788684-8znpr.txt)
[che-codewind-performance-workspace1zycwn4fbqw199tr-864d655fc-jgjd9.txt](https://github.com/eclipse/codewind/files/4389080/che-codewind-performance-workspace1zycwn4fbqw199tr-864d655fc-jgjd9.txt)

**Comment by jagraj** [comment]
With the leatest fix, I executed load test in both "Hybrid" and "Che" environments. 

**1. Hybrid/Remote**

Plugin : Eclipse 
Browser : Chrome
OCP version : 4.3 (Fyre)

I created four projects Default (WebSphere liberty, Spring) and Appsody (Node.js RED, Node.js Scaffold). After WebSphere liberty project started running, I started load test and the request submitted successfully and load test started running. While this test was running then I open another performance window for  Appsody Node.js Scaffold and I got "Offline dialog" which is expected based on latest fix. The message says offline though but it should say another test is in progress and can not be run at this time. I did wait until load run complete for WebSphere liberty project but the dialog never goes away and it always stays as "Offline" even though I click on "Refresh" button. I did close all windows and opened new "Performance dashboard" window for Appsody Node.js Scaffold and I always get "Offline" dialog. 

Here are my screenshots and logs for this issue. 

![image](https://user-images.githubusercontent.com/713849/77684635-9e3c6580-6f70-11ea-94de-b43b9a1a7364.png)


[gate-keeper.txt](https://github.com/eclipse/codewind/files/4389230/gate-keeper.txt)
[performance.txt](https://github.com/eclipse/codewind/files/4389231/performance.txt)
[pfe.txt](https://github.com/eclipse/codewind/files/4389232/pfe.txt)


**2. Eclipse che** 

Browser : Firefox
OCP version : 4.3 (Fyre)

I see same behavior as Sujeily reported in the previous comment, the status get stuck as "Requested" and I do not see "Offline" dialog at all in Che scenario and the status says always "Connected". The other issue we noticed that we are not getting results back for the load test in che. 

![image](https://user-images.githubusercontent.com/713849/77684132-e1e29f80-6f6f-11ea-84f5-6294bca3fa94.png)

![image](https://user-images.githubusercontent.com/713849/77684271-1d7d6980-6f70-11ea-8839-6a734a1f78dc.png)

[performance.txt](https://github.com/eclipse/codewind/files/4389222/performance.txt)
[pfe.txt](https://github.com/eclipse/codewind/files/4389224/pfe.txt)

**Comment by jagraj** [comment]
@elsony 
Based on our conversation, can someone from your team try this latest fix in your environments as well.?

**Comment by elsony** [comment]
yes, we are testing it out. We'll post the result when available.

**Comment by tobespc** [comment]
ok, so what happens if you don't try and start  multiple projects at once and as soon as one is ready fire load through it ?   What do you get if you wait for however many projects you choose to start to all be in running state before attempting the first load test ?  We want to make sure we are testing a realistic development scenario and not hitting errors due to unrealistic stress testing.  The use case we are testing here is for a developer to start up codewind, straight away start multiple projects building and running and then as soon as one of those is ready, fire load at it and then swap back to other projects and fire load at those.   As I said earlier, performance testing multiple projects concurrently is not what we consider to be the golden path for a developer, its more of an edge scenario.

**Comment by jgwest** [comment]
Codewind 0.10.0 on Che, on generic Kubernetes, with Chrome, never load testing more than one application at a time:
- No problems one-at-a-time load-testing a Node application or  Microprofile application (at least based on my familiarity with this fxnality, but see below)
- When I disconnected my machine from the cluster (via disconnecting from VPN), I successfully got the new 'Dashboard Offline' dialog. When I reconnected, the dialog eventually went away without needing to refresh.
- Not able to reproduce the Che behaviour that Sujeily described here (https://github.com/eclipse/codewind/issues/2284#issuecomment-604599866)
- I always got results back from Che.

But, not sure if this is expected behaviour:
- If I start a test, and then refresh that page, should I see the status of that test? Because I never did.
- When I started a load test on MP, then immediately started a load test on Node, the Node one was stuck at 'Requested...'. The status on the Node one did not change after MP finished, and it wasn't clear if it was just not reporting the status, or if it was actually not load testing.
- At one point it *seemed* like I could not run any more load tests, because I always got that 'a load test is running' error back.

**Comment by DavidG1011** [comment]
My findings with Che on OKD 4.2:

1. Created one of each: MicroProfile, Swift, Spring, and Nodejs. Waited for all of them to reach Running, Build succeeded status.

2. Ran a load test with the MicroProfile project. Test completed successfully and result was shown. After a brief pause, I started another load test, and the load test was stuck in the "Requested" state.

3. Attempted to cancel the load test, but that appeared to get stuck as well.

4. Deleted all 4 projects.

5. Recreated 1 MicroProfile project. Waited for it to reach Running, Build succeeded status.

6. Ran a load test. Load test got stuck in "Requested" state.


[pfe.txt](https://github.com/eclipse/codewind/files/4389892/pfe.txt)
[performance.txt](https://github.com/eclipse/codewind/files/4389894/performance.txt)

**Comment by tobespc** [comment]
found a possible culprit in the code..... investigating....

**Comment by sujeilyfonseca** [comment]
I retested this issue with the latest Codewind 0.10.0. I covered scenarios with Eclipse Che, using a Fyre OCP 4.3 cluster and a Google Chrome browser on Mac; and VS Code (Remote) using a Windows VM and a Firefox browser. 

**1. VS Code (Remote):**
I cleaned my environment. I proceeded to install the latest Codewind 0.10.0. After that, I created a new remote instance with the newest `cwctl`. Then, I created four default projects and waited for the projects to be in "Running" state.

<img width="1676" alt="Screen Shot 2020-03-27 at 9 01 56 AM" src="https://user-images.githubusercontent.com/49531960/77766814-8a4b3f00-7016-11ea-96cb-59b8c770134a.png">

The “Offline” modal window was present for every project. It didn't get away even after waiting for a while. I was not able to run any load test.

<img width="1654" alt="Screen Shot 2020-03-27 at 9 02 06 AM" src="https://user-images.githubusercontent.com/49531960/77765157-36d7f180-7014-11ea-994d-7578cc113dca.png">

<img width="1654" alt="Screen Shot 2020-03-27 at 9 06 53 AM" src="https://user-images.githubusercontent.com/49531960/77765161-38a1b500-7014-11ea-8b95-c0a99793ef81.png">

<img width="1655" alt="Screen Shot 2020-03-27 at 9 07 41 AM" src="https://user-images.githubusercontent.com/49531960/77766570-35a7c400-7016-11ea-92dc-0878cdfd930a.png">

<img width="1657" alt="Screen Shot 2020-03-27 at 9 07 17 AM" src="https://user-images.githubusercontent.com/49531960/77766576-380a1e00-7016-11ea-98fa-76ed7154209f.png">

Refreshing the browser didn't work.

</br> 

**2. Eclipse Che:**
I deployed a new Codewind 0.10.0 workspace and proceeded to create four default projects and waited for the projects to be in "Running" state.

<img width="1676" alt="Screen Shot 2020-03-27 at 9 22 30 AM" src="https://user-images.githubusercontent.com/49531960/77766845-933c1080-7016-11ea-9804-baa651e01f14.png">

First, I executed a load test with a Node project, and I successfully retrieved the load test results. When that test was in progress, I was not allowed to run a new request with another project, which is good. I received the error message saying that another test was in progress. I only tested this to ensure the fix was working. Then, I closed that tab and continued testing single projects/tabs.

<img width="1668" alt="Screen Shot 2020-03-27 at 9 20 12 AM" src="https://user-images.githubusercontent.com/49531960/77766976-c5e60900-7016-11ea-80ea-99a32e01ac80.png">

<img width="1598" alt="Screen Shot 2020-03-27 at 9 04 24 AM" src="https://user-images.githubusercontent.com/49531960/77767008-d1d1cb00-7016-11ea-8a09-638d7ad4d740.png">

After getting test results with the Node project, I executed a load test with a Spring project, and I successfully retrieved test results.

<img width="1672" alt="Screen Shot 2020-03-27 at 9 00 26 AM" src="https://user-images.githubusercontent.com/49531960/77767049-ddbd8d00-7016-11ea-8f74-7bee0532700f.png">

Then, I tried a MicroProfile project, and I didn’t get the load test results after waiting for the load test to finish. From now on, every project presents the error saying that another test is in progress.

<img width="1668" alt="Screen Shot 2020-03-27 at 9 20 12 AM" src="https://user-images.githubusercontent.com/49531960/77767078-e8782200-7016-11ea-80b7-0c31f3cf334b.png">

<img width="1672" alt="Screen Shot 2020-03-27 at 9 09 53 AM" src="https://user-images.githubusercontent.com/49531960/77767093-eca43f80-7016-11ea-897c-cacdc2aa07e6.png">

<img width="1663" alt="Screen Shot 2020-03-27 at 9 08 59 AM" src="https://user-images.githubusercontent.com/49531960/77767094-ed3cd600-7016-11ea-98c6-ce5131d11ca5.png">

</br>

**Logs:**
[vscode-remote-codewind-pfe-k8a6kny4-66d7897c58-rtpzv.txt](https://github.com/eclipse/codewind/files/4394106/vscode-remote-codewind-pfe-k8a6kny4-66d7897c58-rtpzv.txt

[che-codewind-workspace1jvpx4kuou075wqs-79c856dbc6-tlw6n.txt](https://github.com/eclipse/codewind/files/4394070/che-codewind-workspace1jvpx4kuou075wqs-79c856dbc6-tlw6n.txt)
[che-codewind-performance-workspace1jvpx4kuou075wqs-797c5764ff-pw2jk.txt](https://github.com/eclipse/codewind/files/4394071/che-codewind-performance-workspace1jvpx4kuou075wqs-797c5764ff-pw2jk.txt)

</br>
**Note:** Something is not right with the Websphere Liberty MicroProfile stack since I have not retrieved test results with that stack more than once. I can reproduce this behavior locally, which let me think this can be a separate issue.

However, if the user didn't retrieve results with a project, this will block subsequent tests. Previously, subsequent tests were blocked with the status "Requested" and now with an error indicating that another test is in progress.

**Comment by jagraj** [comment]
**Hybrid/Remote:**

With the latest fix without "Offline" dialog, we see some improvements better than last fix. Looks like WebSphere Liberty and Open Liberty projects load tests failing and not getting load test results and I also injected metrics explicitly to see if I can get results but I did not get results. When these projects fail and I no longer submit load tests for any other projects and it always says another load test is in progress.



[gate-keeper.txt](https://github.com/eclipse/codewind/files/4405910/gate-keeper.txt)
[performance.txt](https://github.com/eclipse/codewind/files/4405911/performance.txt)
[pfe.txt](https://github.com/eclipse/codewind/files/4405912/pfe.txt)

**local**

On local also, WebSphere and Open Liberty projects does not produce results if we run load tests from performance dashboard.

![image](https://user-images.githubusercontent.com/713849/77971594-15e5f980-72bd-11ea-8341-76f1a6676062.png)

![image](https://user-images.githubusercontent.com/713849/77971641-3615b880-72bd-11ea-8e30-186d6c4e218a.png)


[performance.txt](https://github.com/eclipse/codewind/files/4405932/performance.txt)
[pfe.txt](https://github.com/eclipse/codewind/files/4405933/pfe.txt)

**Comment by sujeilyfonseca** [comment]
I retested this issue with the latest Codewind 0.10.0. I covered scenarios with Eclipse Che, using a Fyre OCP 4.3 cluster and a Google Chrome browser on Mac; and VS Code (Remote) using a Windows VM and a Firefox browser. I see some improvements with the latest fix.

**1. VS Code (Remote)**
I cleaned my environment. I proceeded to install the latest Codewind 0.10.0. After that, I created a new remote instance with the newest `cwctl`. Then, I created five default projects and waited for the projects to be in "Running" state.

<img width="1676" alt="Screen Shot 2020-03-30 at 7 27 14 PM" src="https://user-images.githubusercontent.com/49531960/77973482-572cd800-72c2-11ea-9614-d515dcd4b00b.png">

My first load test with a Spring project was stuck in "Requested". The status never changed, but after some time, I obtained test results.

![Screen Shot 2020-03-30 at 7 26 51 PM](https://user-images.githubusercontent.com/49531960/77973106-4e87d200-72c1-11ea-8e52-e23dbfa85096.png)

![Screen Shot 2020-03-30 at 7 30 20 PM](https://user-images.githubusercontent.com/49531960/77973127-59dafd80-72c1-11ea-95ad-755144ff61aa.png)

After that, I executed additional load tests, and the status changed as expected. With these tests, I successfully retrieved results. When a test was in progress, I was not allowed to run a new request with another project, which is expected.

![Screen Shot 2020-03-30 at 7 37 18 PM](https://user-images.githubusercontent.com/49531960/77974048-061de380-72c4-11ea-906b-56da7de63194.png)

I didn't obtain load test results with an Open Liberty project, but this didn't block subsequent tests.

![Screen Shot 2020-03-30 at 7 42 41 PM](https://user-images.githubusercontent.com/49531960/77973825-695b4600-72c3-11ea-8b92-3e349071889e.png)

MicroProfile projects don't produce test results and block subsequent tests.

![Screen Shot 2020-03-30 at 7 46 41 PM](https://user-images.githubusercontent.com/49531960/77973776-44ff6980-72c3-11ea-98a4-6c38130e239a.png)

![Screen Shot 2020-03-30 at 7 46 52 PM](https://user-images.githubusercontent.com/49531960/77973779-47fa5a00-72c3-11ea-8ded-96ed9e1a50d4.png)

![Screen Shot 2020-03-30 at 7 47 22 PM](https://user-images.githubusercontent.com/49531960/77973785-4a5cb400-72c3-11ea-8a02-7f6e46229e03.png)

</br>

**2. Eclipse Che**
I deployed a new Codewind 0.10.0 workspace and proceeded to create five default projects and waited for the projects to be in "Running" state.

<img width="1674" alt="Screen Shot 2020-03-30 at 8 19 03 PM" src="https://user-images.githubusercontent.com/49531960/77973958-ba6b3a00-72c3-11ea-88e6-28c861adda65.png">

I successfully executed and retrieved test results with some projects. When a test was in progress, I was not allowed to run a new request with another project, which is expected.

<img width="1655" alt="Screen Shot 2020-03-30 at 7 29 52 PM" src="https://user-images.githubusercontent.com/49531960/77973626-c1457d00-72c2-11ea-9f21-ac2ea21e44e0.png">

<img width="1652" alt="Screen Shot 2020-03-30 at 8 24 03 PM" src="https://user-images.githubusercontent.com/49531960/77974207-7167b580-72c4-11ea-98f6-8e2e573eb429.png">

I didn't obtain load test results with an Open Liberty project, but this didn't block subsequent tests.

<img width="1670" alt="Screen Shot 2020-03-30 at 7 44 37 PM" src="https://user-images.githubusercontent.com/49531960/77973908-960f5d80-72c3-11ea-92ca-e1734c3c3b33.png">

However, MicroProfile projects don't produce test results and block subsequent tests.

<img width="1676" alt="Screen Shot 2020-03-30 at 7 48 27 PM" src="https://user-images.githubusercontent.com/49531960/77973894-8c85f580-72c3-11ea-9490-bbb098e1bff5.png">

<img width="1667" alt="Screen Shot 2020-03-30 at 7 48 35 PM" src="https://user-images.githubusercontent.com/49531960/77973897-8f80e600-72c3-11ea-88ad-8115ca223b9c.png">

<img width="1667" alt="Screen Shot 2020-03-30 at 7 48 49 PM" src="https://user-images.githubusercontent.com/49531960/77973900-914aa980-72c3-11ea-9a3b-90bef8b4c93e.png">


</br>

**Logs:**
[vscode-remote-codewind-pfe-k8f393kl-64cbc5c8cf-22zxlc.txt](https://github.com/eclipse/codewind/files/4406002/vscode-remote-codewind-pfe-k8f393kl-64cbc5c8cf-22zxlc.txt)
[vscode-remote-codewind-performance-k8f393kl-76f8756dd7-6dd2b.txt](https://github.com/eclipse/codewind/files/4406003/vscode-remote-codewind-performance-k8f393kl-76f8756dd7-6dd2b.txt)

[che-codewind-workspace9vgevotcwcfimi5q-6dbb9f6cbd-bl4wh.txt](https://github.com/eclipse/codewind/files/4406004/che-codewind-workspace9vgevotcwcfimi5q-6dbb9f6cbd-bl4wh.txt)
[che-codewind-performance-workspace9vgevotcwcfimi5q-7869cd6dff-tg94j.txt](https://github.com/eclipse/codewind/files/4406006/che-codewind-performance-workspace9vgevotcwcfimi5q-7869cd6dff-tg94j.txt)

</br>

**Note:** The default Open Liberty and WebSphere Liberty MicroProfile projects are not producing test results in local scenarios either.

**Comment by DavidG1011** [comment]
My test results are similar to Jag's.

**Che:**

I was able to run a few load tests successfully with Java Spring, Nodejs, and Swift projects. I then tested out a load test with a Java Microprofile project. The test did not produce any results, and I was unable to run any subsequent load tests.

Test running:

<img width="1680" alt="Screen Shot 2020-03-30 at 8 03 49 PM" src="https://user-images.githubusercontent.com/30476587/77973724-12557100-72c3-11ea-9261-58b167ce5140.png">


Subsequent load test:

<img width="1680" alt="Screen Shot 2020-03-30 at 8 04 12 PM" src="https://user-images.githubusercontent.com/30476587/77973728-197c7f00-72c3-11ea-8175-2e1fb7b6dfbe.png">

[che-performance.txt](https://github.com/eclipse/codewind/files/4406074/che-performance.txt)
[che-pfe.txt](https://github.com/eclipse/codewind/files/4406075/che-pfe.txt)


**Local:**

I then tried running load tests locally and saw similar results. 

Test ran the first time for Microprofile:

<img width="1680" alt="Screen Shot 2020-03-30 at 7 58 17 PM" src="https://user-images.githubusercontent.com/30476587/77973836-6f512700-72c3-11ea-9ba4-9f530023d5b5.png">

Subsequent:

<img width="1680" alt="Screen Shot 2020-03-30 at 7 59 12 PM" src="https://user-images.githubusercontent.com/30476587/77973870-8132ca00-72c3-11ea-95bc-1f2fa5aefef2.png">


I then attempted to run a load test for a Java Open Liberty type, and was able to, but with the same result as the Microprofile load test; received no results, and was unable to run subsequent load tests.

[local_pfe.txt](https://github.com/eclipse/codewind/files/4406072/local_pfe.txt)
[local_performance.txt](https://github.com/eclipse/codewind/files/4406073/local_performance.txt)

**Comment by tobespc** [comment]
you are all seeing different results to me :-(

**Comment by tobespc** [comment]
current situation is as follows

For microprofile I get the following
  * Load will run for specified time
  * metrics.json is created and written to codewind-workspace/projname/load-test/xxxxxxx/metrics.json
  * the .hcd file does get created in the application container (/home/default/app/load-test/xxxxx/healthcenterxxxx.hcd) but only when the project status (top left of perf dash) goes to 'completed' state
  * at this point, we have exhausted the 20 tries to get this file
  * I have a findHCD branch that contains changes (not working) to try and detect the name of the hcd file created in the app container, and then copy that straight into the pfe container

**Comment by tobespc** [comment]
Fix is in against 0.11 and I'll make the same fix to master.   Further changes to master can be tracked in a separate issue

**Comment by jagraj** [comment]
@tobespc @markcor11 
With the latest 0.11 images, some how the message we used to display to the users saying another load test is in progress does not appear any more. This message was blocking user for not to submit another request before and not now. After injecting metrics for OpenLiberty on che I was able to run the load the test. When I tried to submit another request when the load was running for another test and that one get stuck into "Requested" state for ever. I think we still need some improvements. 

![image](https://user-images.githubusercontent.com/713849/78715072-dc2f7700-78ea-11ea-9f96-6f394c9f9db7.png)

Logs:
[performance.txt](https://github.com/eclipse/codewind/files/4446811/performance.txt)
[pfe.txt](https://github.com/eclipse/codewind/files/4446812/pfe.txt)

**Comment by tobespc** [comment]
@jagraj , ok, will continue working this morning on it

**Comment by jagraj** [comment]
@tobespc @markcor11 
I tried latest fixes, I see some improvement. Now to we do show message to the user that "Another load test is in progress" when another test is in progress. The first test load run completed and went into "Running" state and did not come back to "Ready" state to submit another load test and it always complains another load test is in progress. I am also noticing this error intermittently when I try to access performance dash board "NotFoundError"  404 Endpoint Get /appmetrics-dash not found. Here are all the screenshots and logs from these tests. Even after test completed, the status stays as "Running".

<img width="1674" alt="Screen Shot 2020-04-09 at 12 39 44 AM" src="https://user-images.githubusercontent.com/713849/78859427-49333180-79fd-11ea-9c32-6e11584e254b.png">
<img width="1680" alt="Screen Shot 2020-04-09 at 12 41 18 AM" src="https://user-images.githubusercontent.com/713849/78859428-49cbc800-79fd-11ea-9b5a-df020f52467f.png">
<img width="1669" alt="Screen Shot 2020-04-09 at 12 32 50 AM" src="https://user-images.githubusercontent.com/713849/78859429-49cbc800-79fd-11ea-93e6-964305db6ab8.png">
<img width="1678" alt="Screen Shot 2020-04-09 at 12 30 54 AM" src="https://user-images.githubusercontent.com/713849/78859430-49cbc800-79fd-11ea-8af9-e6f9b134078f.png">
<img width="1679" alt="Screen Shot 2020-04-09 at 12 28 34 AM" src="https://user-images.githubusercontent.com/713849/78859431-4a645e80-79fd-11ea-96b1-29d57e94701a.png">

**Comment by tobespc** [comment]
This issue is becoming a bit of a mess now, so let me try and clarify how things are working.

The original stop ship of being stuck in requested state has been resolved and as such I would like us to consider removing the stop-ship label, closing this item and having a new issue with just the current failing behaviour documented for discussion

I believe the behaviour you are seeing is limited to Microprofile and OpenLiberty style projects now.

For issues of appearing to be stuck in 'Running' state whilst the test has finished, this is how the messages currently work
1) load test is run and after a few changes the state goes to Running (xx) with a timer
2) When the time ends, that indicates the load has finished but the performance dashboard is still busy collecting the information generated.  Thats why it stays in Running state (and it can stay there for quite some time).  We have plans to change that in 0.12

Running load and collecting metrics is not an instance thing, a lot of processing is happening once the load finishes and we should bear this in mind (and make it clearer to Jane) that for some project types (mainly Java) this can take time

For the 404 error , that implies that the application is not running or the endpoint of apmetrics-dash is incorrect.  How are you getting to page ?  Thats a node project you are trying but is that an appsody Node project or codewind template ?  Remember, the appsody node project must be including appmetrics module to work.   This is a new issue and so can this be raised as such please

**Comment by tobespc** [comment]
Change made to make it clear through the status when the collecting of data is taking place

**Comment by jagraj** [comment]
@tobespc @markcor11 
Thanks for the fix, We are going to get latest 0.11 images and give ot a try to see if we can reproduce this "Requested" state stuck issue. If we are able to reproduce this issue, then we will close this issue and open new issues for any other problems.  Once again thanks for all your support here.

**Comment by sujeilyfonseca** [comment]
Thank you very much for working to solve this issue, @tobespc and @markcor11! I appreciate all your help!

I verified this issue with the latest Codewind 0.11.0 in local and cloud-hosted scenarios. The original problem was resolved, but there is an issue with remote where the test first can present a status stuck in "Requested" in the first try, so if the user sees that, the refresh will make it work. I'll create another issue for the documentation needed.

Regarding the other issues with the Performance Dashboard, I'll proceed to open a separate issue to address different problems.
